storage/index/block.go

```diff
@@ -0,0 +1,377 @@
+// THE SOFTWARE.
+
+package index
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3ninx/doc"
+	m3ninxindex "github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/index/segment/mem"
+	"github.com/m3db/m3ninx/postings"
+	"github.com/m3db/m3ninx/search"
+	"github.com/m3db/m3ninx/search/executor"
+	"github.com/m3db/m3x/context"
+	xerrors "github.com/m3db/m3x/errors"
+	xlog "github.com/m3db/m3x/log"
+)
+
+var (
+	errUnableToWriteBlockClosed     = errors.New("unable to write, index block is closed")
+	errUnableToWriteBlockSealed     = errors.New("unable to write, index block is sealed")
+	errUnableToQueryBlockClosed     = errors.New("unable to query, index block is closed")
+	errUnableToBootstrapBlockClosed = errors.New("unable to bootstrap, block is closed")
+	errUnableToTickBlockClosed      = errors.New("unable to tick, block is closed")
+	errBlockAlreadyClosed           = errors.New("unable to close, block already closed")
+
+	errUnableToSealBlockIllegalStateFmtString  = "unable to seal, index block state: %v"
+	errUnableToWriteBlockUnknownStateFmtString = "[invariant violated] unable to write, unknown index block state: %v"
+)
+
+type blockState byte
+
+const (
+	blockStateClosed blockState = iota
+	blockStateOpen
+	blockStateSealed
+)
+
+type newExecutorFn func() (search.Executor, error)
+
+type block struct {
+	sync.RWMutex
+	state blockState
+
+	bootstrappedSegments []segment.Segment
+	segment              segment.MutableSegment
+
+	newExecutorFn newExecutorFn
+	startTime     time.Time
+	endTime       time.Time
+	opts          Options
+	logger        xlog.Logger
+}
+
+// NewBlock returns a new Block, representing a complete reverse index for the
+// duration of time specified. It is backed by one or more segments.
+func NewBlock(
+	startTime time.Time,
+	blockSize time.Duration,
+	opts Options,
+) (Block, error) {
+	// FOLLOWUP(prateek): use this to track segments when we have multiple segments in a Block.
+	postingsOffset := postings.ID(0)
+	seg, err := mem.NewSegment(postingsOffset, opts.MemSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	b := &block{
+		state:   blockStateOpen,
+		segment: seg,
+
+		startTime: startTime,
+		endTime:   startTime.Add(blockSize),
+		opts:      opts,
+		logger:    opts.InstrumentOptions().Logger(),
+	}
+	b.newExecutorFn = b.executorWithRLock
+
+	return b, nil
+}
+
+func (b *block) StartTime() time.Time {
+	return b.startTime
+}
+
+func (b *block) EndTime() time.Time {
+	return b.endTime
+}
+
+func (b *block) WriteBatch(inserts []WriteBatchEntry) (WriteBatchResult, error) {
+	b.RLock()
+	defer b.RUnlock()
+	if b.state != blockStateOpen {
+		// NB: releasing all references to inserts
+		for _, insert := range inserts {
+			insert.OnIndexSeries.OnIndexFinalize()
+		}
+		return WriteBatchResult{
+			NumError: int64(len(inserts)),
+		}, writeBatchErrorInvalidState(b.state)
+	}
+
+	// FOLLOWUP(prateek): docs array pooling
```

Comment:
Hm, since writes to the index are batched you could take a write lock in this method and cache the slice on the index block itself?  Can avoid pooling that way.

Reply:
Definitely. That was the plan but didn't do it cause this code used to live in the index until this PR. Will make the change now.

Reply:
done